### PR TITLE
#BN-30 feat: logout from the App

### PR DESCRIPTION
**Pull request by nellykawera**
# PR Description
Please include a brief description of the pull request you want to raise.
The admin, vendor and buyer should be able to logout
# Description of tasks that were expected to be completed
List all the tasks and sub-tasks that were assigned to you Make sure you check every completed task. list them in this style
- [x] As an authenticated user, I want to be able to logout, so that none else can access my account as well
# Description of tasks that were expected to be completed and that are not yet completed
- [x ] Authenticated user should be able to logout.
- [x ] Only authenticated user should be able to access protected resourses/URLs
- [x ] Appropriate error messages are displayed on unsuccessful attempts
# How this may be tested?
Please describe the tests that you ran to verify your functionalities. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
1. run `npm i` to install all required dependencies
2. run `npm run dev`
3. swagGer UI "DELETE"
# Screenshots (If appropriate/applicable)
Provide screenshots related to your PR in case it's appropriate and applicable
# Please check this Checklist before you submit your PR:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My code generates no warnings
- [x] There are no vulnerabilities

### CODE REVIEW

**Review by uwdjany — COMMENTED** [review]
well done!

**Review by uwdjany — DISMISSED** [review]

File: src/locales/fr/translation.json [hunk 1]
```diff
@@ -1,3 +1,9 @@
 {
     "welcome_message": "Bienvenue dans l'API de l'application Ecommerce de Cogito"
+    "loginUser_message": "Veuillez fournir un e-mail et un mot de passe pour vous connecter!"
+    "createUser_message": "Veuillez fournir le prénom, le nom, l'e-mail, le mot de passe, le téléphone et le rôle pour créer un utilisateur!"
+<<<<<<< HEAD
+     "logoutUser_message": "L'utilisateur s'est déconnecté avec succès"
+=======
+>>>>>>> 67efa140f34198bc0baf99d91940080127fad01a
```
Comment (chrisostome01) [inline src/locales/fr/translation.json]: Is this supposed to be here

File: tests.js/testUser.test.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable indent */
```
Comment (chrisostome01) [inline tests.js/testUser.test.js]: Fix the linting issues instead of ignoring them

File: tests.js/testUser.test.js
Comment (chrisostome01) [inline tests.js/testUser.test.js:1]: Consider formatting this file with prettier eslint

File: src/controllers/UserController.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    });
+  }
+}
+
+export async function logoutUser(req, res) {
+  // Perform necessary actions for logging out, such as invalidating the session, removing tokens.
+  // Example:
+  // await invalidateSession(req.session);
+  // await removeTokens(req.user);
+
```
Comment (chrisostome01) [inline src/controllers/UserController.js]: The logout functionality is not really working.
Uncomment these comments and make sure to handle errors

File: src/controllers/UserController.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  const {
+    firstName, lastName, email, password, phone, role,
+  } = req.body;
+  if (!firstName || !lastName || !email || !password || !phone || !role) {
+    return res.status(400).json({
+      status: 400,
+      message:
+        'Please provide firstName, lastName, email, password, phone, and role to create a user!',
+    });
+  }
```
Comment (chrisostome01) [inline src/controllers/UserController.js]: Use Joi for this validation

File: src/controllers/UserController.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    where: {
+      email,
+    },
+  });
+  if (emailExists) {
+    return res.status(409).json({
+      status: 400,
+      message: 'An account with that email already exists!',
+    });
+  }
```
Comment (chrisostome01) [inline src/controllers/UserController.js]: I would recommend wrapping this inside try and catch block

**Review by chrisostome01 — CHANGES_REQUESTED** [review]

File: tests.js/testUser.test.js
Comment (chrisostome01) [inline tests.js/testUser.test.js:1]: Why is this file is deleted

File: src/controllers/UserController.js
Comment (chrisostome01) [inline src/controllers/UserController.js:1]: let us use one file for the user controller and I would take an example from how other controllers are being named and stick to the same casing..

File: src/controllers/userController.js [hunk 6]
```diff
@@ -276,3 +279,9 @@ export async function deleteUser(req, res) {
     });
   }
 }
+export async function logoutUser(req, res) {
+  return res.status(200).json({
+    status: 200,
+    message: 'User logged out successfully',
+  });
```
Comment (chrisostome01) [inline src/controllers/userController.js]: I think this is not enough to logout the user, would you try to improve this process?

**Review by chrisostome01 — CHANGES_REQUESTED** [review]

**Review by dushimeemma — COMMENTED** [review]
Write the unit test for your endpoint please, otherwise good job!

File: src/tests/testUsers.test.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      expect(res.status).to.equal(200);
+      expect(res.body.message).to.equal('User logged in successfully');
+      expect(res.body.token).to.be.a('string');
+    });
+  });
+
+  after(async () => {
+    await User.destroy({ truncate: true });
+  });
+});
```
Comment (dushimeemma) [inline src/tests/testUsers.test.js:45]: Add logout functionality unit test

**Review by dushimeemma — CHANGES_REQUESTED** [review]

File: src/tests/testUsers.test.js (hunk 7 above)
Comment (nellykawera) [reply src/tests/testUsers.test.js:45]: I did thanks

File: src/controllers/userController.js [hunk 8]
```diff
@@ -84,8 +84,180 @@ export async function createUser(req, res) {
...
+// Logout
+export async function logoutUser(req, res) {
+  try {
+    // Clear the session
+    req.session.destroy((err) => {
+      if (err) {
+        throw new Error('Failed to destroy session');
+      }
+      res.clearCookie('sessionID');
+      res.status(200).json({ message: 'Logged out successfully' });
```
Comment (chrisostome01) [inline src/controllers/userController.js]: I think this message would be better if it is translated....

**Review by chrisostome01 — DISMISSED** [review]

File: src/database/migrations/20220409220608-create-user.js [hunk 9]
```diff
@@ -50,7 +50,11 @@ module.exports = {
       },
       status: {
         type: Sequelize.STRING,
+<<<<<<< HEAD:src/database/migrations/20230409220608-create-user.js
+        allowNull: true,
+=======
```
Comment (dushimeemma) [inline src/database/migrations/20220409220608-create-user.js]: what is the meaning of this line?

File: src/database/migrations/20220409220608-create-user.js [hunk 10]
```diff
@@ -50,7 +50,11 @@ module.exports = {
...
         allowNull: false,
+>>>>>>> 61173629cdf84ae40840eb4a389e2058bf3dedb9:src/database/migrations/20220409220608-create-user.js
```
Comment (dushimeemma) [inline src/database/migrations/20220409220608-create-user.js]: what is the meaning of this line?

File: src/database/config/config.json
Comment (dushimeemma) [inline src/database/config/config.json:1]: the config file should be a js file

**Review by dushimeemma — CHANGES_REQUESTED** [review]

File: src/database/config/config.json
Comment (dushimeemma) [inline src/database/config/config.json:1]: Database config is a JS file @nellykawera please do rebase correctly.
`git pull --rebase origin develop`

**Comment by dushimeemma** [comment]
> A deploy for your [Render](https://render.com) PR Server at https://cogito-api-pr-32.onrender.com just failed.
> 
> View details on your dashboard at https://dashboard.render.com/web/srv-cgv835hmbg54u99g37h0.

@nellykawera kindly check the reason why render is failing to deploy your branch

File: src/tests/testUsers.test.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+describe('POST /auth/logout', () => {
+  it('should log out an existing user', async () => {
+    const res = await chai
+      .request(app)
+      .post('/logout')
+      .set({ Authorization: `Bearer ${token}` })
+      .send({
+        email: 'johndoe@example.com',
+        password: 'password',
+      });
```
Comment (Honore123) [inline src/tests/testUsers.test.js:42]: Add expected result

**Review by Honore123 — APPROVED** [review]
🚀

**Review by dushimeemma — APPROVED** [review]
Good job 👍

**Review by uwdjany — APPROVED** [review]

**Pull request merged**